File: src/net/HttpTeamService.cpp
#include "HttpTeamService.h"

#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>

namespace {

ServerUser userFromJson(const QJsonObject &obj)
{
    ServerUser user;
    user.id = obj.value(QStringLiteral("id")).toInt();
    user.email = obj.value(QStringLiteral("email")).toString();
    user.displayName = obj.value(QStringLiteral("displayName")).toString();
    user.isAdmin = obj.value(QStringLiteral("isAdmin")).toBool();
    return user;
}

ProjectMember memberFromJson(const QJsonObject &obj)
{
    ProjectMember member;
    member.userId = obj.value(QStringLiteral("userId")).toInt();
    member.email = obj.value(QStringLiteral("email")).toString();
    member.displayName = obj.value(QStringLiteral("displayName")).toString();
    member.role = obj.value(QStringLiteral("role")).toString();
    return member;
}

QString readErrorMessage(const QJsonDocument &doc, const QString &fallback)
{
    if (!doc.isObject())
        return fallback;
    const QJsonObject obj = doc.object();
    if (obj.contains(QStringLiteral("error")))
        return obj.value(QStringLiteral("error")).toString(fallback);
    if (obj.contains(QStringLiteral("message")))
        return obj.value(QStringLiteral("message")).toString(fallback);
    return fallback;
}

} // namespace

HttpTeamService::HttpTeamService(ApiClient &client)
    : m_client(client)
{
}

bool HttpTeamService::fetchCurrentUser(ServerUser *user) const
{
    int status = 0;
    const QJsonDocument doc = m_client.get(QStringLiteral("/api/v1/auth/me"), &status);
    if (status != 200 || !doc.isObject()) {
        m_lastError = readErrorMessage(doc, m_client.lastError());
        return false;
    }
    if (user)
        *user = userFromJson(doc.object());
    return true;
}

QList<ServerUser> HttpTeamService::listUsers() const
{
    int status = 0;
    const QJsonDocument doc = m_client.get(QStringLiteral("/api/v1/admin/users"), &status);
    if (status != 200 || !doc.isArray()) {
        m_lastError = readErrorMessage(doc, m_client.lastError());
        return {};
    }

    QList<ServerUser> users;
    for (const QJsonValue &value : doc.array()) {
        if (value.isObject())
            users.append(userFromJson(value.toObject()));
    }
    return users;
}

ServerUser HttpTeamService::createUser(const QString &email, const QString &displayName,
                                         const QString &password, bool isAdmin) const
{
    QJsonObject body;
    body.insert(QStringLiteral("email"), email.trimmed());
    body.insert(QStringLiteral("displayName"), displayName.trimmed());
    body.insert(QStringLiteral("password"), password);
    body.insert(QStringLiteral("isAdmin"), isAdmin);

    int status = 0;
    const QJsonDocument doc = m_client.post(QStringLiteral("/api/v1/admin/users"), body, &status);
    if (status != 201 || !doc.isObject()) {
        m_lastError = readErrorMessage(doc, m_client.lastError());
        return {};
    }
    return userFromJson(doc.object());
}

QList<ProjectMember> HttpTeamService::listMembers(int projectId) const
{
    int status = 0;
    const QJsonDocument doc = m_client.get(
        QStringLiteral("/api/v1/projects/%1/members").arg(projectId), &status);
    if (status != 200 || !doc.isArray()) {
        m_lastError = readErrorMessage(doc, m_client.lastError());
        return {};
    }

    QList<ProjectMember> members;
    for (const QJsonValue &value : doc.array()) {
        if (value.isObject())
            members.append(memberFromJson(value.toObject()));
    }
    return members;
}

ProjectMember HttpTeamService::addMember(int projectId, const QString &email,
                                         const QString &role) const
{
    QJsonObject body;
    body.insert(QStringLiteral("email"), email.trimmed());
    body.insert(QStringLiteral("role"), role);

    int status = 0;
    const QJsonDocument doc = m_client.post(
        QStringLiteral("/api/v1/projects/%1/members").arg(projectId), body, &status);
    if (status != 201 || !doc.isObject()) {
        m_lastError = readErrorMessage(doc, m_client.lastError());
        return {};
    }
    return memberFromJson(doc.object());
}

ProjectMember HttpTeamService::updateMemberRole(int projectId, int userId,
                                                const QString &role) const
{
    QJsonObject body;
    body.insert(QStringLiteral("role"), role);

    int status = 0;
    const QJsonDocument doc = m_client.patch(
        QStringLiteral("/api/v1/projects/%1/members/%2").arg(projectId).arg(userId), body, &status);
    if (status != 200 || !doc.isObject()) {
        m_lastError = readErrorMessage(doc, m_client.lastError());
        return {};
    }
    return memberFromJson(doc.object());
}

bool HttpTeamService::removeMember(int projectId, int userId) const
{
    int status = 0;
    if (!m_client.del(QStringLiteral("/api/v1/projects/%1/members/%2").arg(projectId).arg(userId),
                      &status)) {
        m_lastError = m_client.lastError();
        return false;
    }
    if (status != 204) {
        m_lastError = QStringLiteral("HTTP %1").arg(status);
        return false;
    }
    return true;
}

QString HttpTeamService::lastError() const
{
    if (!m_lastError.isEmpty())
        return m_lastError;
    return m_client.lastError();
}

File: Delphi/Net/HttpTeamService.pas
unit HttpTeamService;

interface

uses
  System.SysUtils, System.JSON, IsmsDomain, ApiClient, HttpJson;

type
  THttpTeamService = class
  private
    FClient: TApiClient;
    FLastError: string;
    class function ReadErrorMessage(ADoc: TJSONValue; const AFallback: string): string; static;
  public
    constructor Create(AClient: TApiClient);
    function FetchCurrentUser(out AUser: TServerUser): Boolean;
    function ListUsers: TArray<TServerUser>;
    function ListMembers(AProjectId: Integer): TArray<TProjectMember>;
    function AddMember(AProjectId: Integer; const AEmail, ARole: string): TProjectMember;
    function UpdateMemberRole(AProjectId, AUserId: Integer; const ARole: string): TProjectMember;
    function RemoveMember(AProjectId, AUserId: Integer): Boolean;
    function CreateUser(const AEmail, ADisplayName, APassword: string; AIsAdmin: Boolean): TServerUser;
    function GetLastError: string;
    property LastError: string read GetLastError;
  end;

implementation

constructor THttpTeamService.Create(AClient: TApiClient);
begin
  inherited Create;
  FClient := AClient;
end;

class function THttpTeamService.ReadErrorMessage(ADoc: TJSONValue; const AFallback: string): string;
var
  Obj: TJSONObject;
begin
  Result := AFallback;
  if not (ADoc is TJSONObject) then
    Exit;
  Obj := TJSONObject(ADoc);
  if Obj.TryGetValue<string>('error', Result) then
    Exit;
  if Obj.TryGetValue<string>('message', Result) then
    Exit;
  Result := AFallback;
end;

function THttpTeamService.FetchCurrentUser(out AUser: TServerUser): Boolean;
var
  Doc: TJSONValue;
  Status: Integer;
begin
  FillChar(AUser, SizeOf(AUser), 0);
  Doc := FClient.Get('/api/v1/auth/me', Status);
  try
    if (Status <> 200) or not (Doc is TJSONObject) then
    begin
      FLastError := ReadErrorMessage(Doc, FClient.LastError);
      Exit(False);
    end;
    AUser := ServerUserFromJson(TJSONObject(Doc));
    Result := True;
  finally
    Doc.Free;
  end;
end;

function THttpTeamService.GetLastError: string;
begin
  if FLastError <> '' then
    Exit(FLastError);
  Result := FClient.LastError;
end;

function THttpTeamService.ListUsers: TArray<TServerUser>;
var
  Doc: TJSONValue;
  Arr: TJSONArray;
  I: Integer;
  List: TArray<TServerUser>;
  Status: Integer;
begin
  SetLength(List, 0);
  Doc := FClient.Get('/api/v1/admin/users', Status);
  try
    if (Status <> 200) or not (Doc is TJSONArray) then
    begin
      FLastError := ReadErrorMessage(Doc, FClient.LastError);
      Exit;
    end;
    Arr := TJSONArray(Doc);
    SetLength(List, Arr.Count);
    for I := 0 to Arr.Count - 1 do
      if Arr.Items[I] is TJSONObject then
        List[I] := ServerUserFromJson(TJSONObject(Arr.Items[I]));
  finally
    Doc.Free;
  end;
  Result := List;
end;

function THttpTeamService.ListMembers(AProjectId: Integer): TArray<TProjectMember>;
var
  Doc: TJSONValue;
  Arr: TJSONArray;
  I: Integer;
  List: TArray<TProjectMember>;
  Status: Integer;
begin
  SetLength(List, 0);
  Doc := FClient.Get(Format('/api/v1/projects/%d/members', [AProjectId]), Status);
  try
    if (Status <> 200) or not (Doc is TJSONArray) then
    begin
      FLastError := ReadErrorMessage(Doc, FClient.LastError);
      Exit;
    end;
    Arr := TJSONArray(Doc);
    SetLength(List, Arr.Count);
    for I := 0 to Arr.Count - 1 do
      if Arr.Items[I] is TJSONObject then
        List[I] := ProjectMemberFromJson(TJSONObject(Arr.Items[I]));
  finally
    Doc.Free;
  end;
  Result := List;
end;

function THttpTeamService.AddMember(AProjectId: Integer; const AEmail, ARole: string): TProjectMember;
var
  Body: TJSONObject;
  Doc: TJSONValue;
  Status: Integer;
begin
  FillChar(Result, SizeOf(Result), 0);
  Body := TJSONObject.Create;
  try
    Body.AddPair('email', Trim(AEmail));
    Body.AddPair('role', ARole);
    Doc := FClient.PostJson(Format('/api/v1/projects/%d/members', [AProjectId]), Body, Status);
    try
      if (Status <> 201) or not (Doc is TJSONObject) then
      begin
        FLastError := ReadErrorMessage(Doc, FClient.LastError);
        Exit;
      end;
      Result := ProjectMemberFromJson(TJSONObject(Doc));
    finally
      Doc.Free;
    end;
  finally
    Body.Free;
  end;
end;

function THttpTeamService.UpdateMemberRole(AProjectId, AUserId: Integer;
  const ARole: string): TProjectMember;
var
  Body: TJSONObject;
  Doc: TJSONValue;
  Status: Integer;
begin
  FillChar(Result, SizeOf(Result), 0);
  Body := TJSONObject.Create;
  try
    Body.AddPair('role', ARole);
    Doc := FClient.PatchJson(Format('/api/v1/projects/%d/members/%d', [AProjectId, AUserId]), Body, Status);
    try
      if (Status <> 200) or not (Doc is TJSONObject) then
      begin
        FLastError := ReadErrorMessage(Doc, FClient.LastError);
        Exit;
      end;
      Result := ProjectMemberFromJson(TJSONObject(Doc));
    finally
      Doc.Free;
    end;
  finally
    Body.Free;
  end;
end;

function THttpTeamService.RemoveMember(AProjectId, AUserId: Integer): Boolean;
var
  Status: Integer;
begin
  Result := FClient.Delete(Format('/api/v1/projects/%d/members/%d', [AProjectId, AUserId]), Status)
    and (Status = 204);
  if not Result then
    FLastError := FClient.LastError;
end;

function THttpTeamService.CreateUser(const AEmail, ADisplayName, APassword: string;
  AIsAdmin: Boolean): TServerUser;
var
  Body: TJSONObject;
  Doc: TJSONValue;
  Status: Integer;
begin
  FillChar(Result, SizeOf(Result), 0);
  Body := TJSONObject.Create;
  try
    Body.AddPair('email', Trim(AEmail));
    Body.AddPair('displayName', Trim(ADisplayName));
    Body.AddPair('password', APassword);
    Body.AddPair('isAdmin', TJSONBool.Create(AIsAdmin));
    Doc := FClient.PostJson('/api/v1/admin/users', Body, Status);
    try
      if (Status <> 201) or not (Doc is TJSONObject) then
      begin
        FLastError := ReadErrorMessage(Doc, FClient.LastError);
        Exit;
      end;
      Result := ServerUserFromJson(TJSONObject(Doc));
    finally
      Doc.Free;
    end;
  finally
    Body.Free;
  end;
end;

end.
